### PR TITLE
Support rename dataset version

### PR DESCRIPTION
**Pull request by SophieGuo410**

### CODE REVIEW

File: ambry-account/src/main/java/com/github/ambry/account/mysql/DatasetDao.java [hunk 1]
```diff
@@ -1828,4 +2084,9 @@ private Dataset getDatasetHelper(int accountId, int containerId, String accountN
     return executeGetDatasetStatement(getDatasetStatement, accountId, containerId, accountName, containerName,
         datasetName);
   }
+
+  private boolean isAutoIncrVersion(String version) {
+    return LATEST.equals(version) || MAJOR.equals(version) || MINOR.equals(version) || PATCH.equals(version)
```
Comment (justinlin-linkedin) [inline ambry-account/src/main/java/com/github/ambry/account/mysql/DatasetDao.java:2089]: we can do something like this 

```final Set<String> ALL_SEMANTIC_COMPONENTS = new HashSet<>(Arrays.asList(LATEST, MAJOR, MINOR, PATCH, REVISION));
return ALL_SEMANTIC_COMPONENTS.contains(version);```

File: ambry-api/src/main/java/com/github/ambry/account/DatasetVersionRecord.java [hunk 2]
```diff
@@ -102,6 +108,10 @@ public long getCreationTimeMs() {
     return creationTimeMs;
   }
 
+  public String getRenameFrom() {
+    return renameFrom;
```
[the remark this answers is not in the document]
Comment (SophieGuo410) [reply ambry-api/src/main/java/com/github/ambry/account/DatasetVersionRecord.java:112]: Sure. updated.

File: ambry-account/src/main/java/com/github/ambry/account/mysql/DatasetDao.java [hunk 3]
```diff
@@ -365,7 +420,91 @@ private DatasetVersionRecord addDatasetVersionHelper(int accountId, int containe
...
+   * @throws SQLException
+   * @throws AccountServiceException
+   */
+  private void renameDatasetVersionHelper(int accountId, int containerId, String datasetName,
+      Dataset.VersionSchema versionSchema, String sourceVersion, String targetVersion)
+      throws SQLException, AccountServiceException {
+    long sourceVersionValue = getVersionBasedOnSchema(sourceVersion, versionSchema);
+    long targetVersionValue = getVersionBasedOnSchema(targetVersion, versionSchema);
+    String renameFromSourceVersion =
+        getDatasetVersionRenameFromHelper(accountId, containerId, datasetName, sourceVersionValue);
```
[the remark this answers is not in the document]
Comment (SophieGuo410) [reply ambry-account/src/main/java/com/github/ambry/account/mysql/DatasetDao.java:446]: Added.

File: ambry-frontend/src/main/java/com/github/ambry/frontend/GetBlobHandler.java [hunk 4]
```diff
@@ -342,14 +342,23 @@ private String getDatasetVersion(RestRequest restRequest) throws RestServiceExce
         restResponseChannel.setHeader(RestUtils.Headers.TARGET_CONTAINER_NAME, containerName);
         restResponseChannel.setHeader(RestUtils.Headers.TARGET_DATASET_NAME, datasetName);
         restResponseChannel.setHeader(RestUtils.Headers.TARGET_DATASET_VERSION, datasetVersionRecord.getVersion());
+        if (datasetVersionRecord.getRenameFrom() != null) {
+          restResponseChannel.setHeader(RestUtils.Headers.TARGET_ORIGINAL_NAMED_BLOB_NAME,
+              datasetVersionRecord.getRenameFrom());
```
Comment (justinlin-linkedin) [inline ambry-frontend/src/main/java/com/github/ambry/frontend/GetBlobHandler.java:347]: why do we need this one?

File: ambry-frontend/src/main/java/com/github/ambry/frontend/GetBlobHandler.java [hunk 5]
```diff
@@ -342,14 +342,23 @@ private String getDatasetVersion(RestRequest restRequest) throws RestServiceExce
         restResponseChannel.setHeader(RestUtils.Headers.TARGET_CONTAINER_NAME, containerName);
         restResponseChannel.setHeader(RestUtils.Headers.TARGET_DATASET_NAME, datasetName);
         restResponseChannel.setHeader(RestUtils.Headers.TARGET_DATASET_VERSION, datasetVersionRecord.getVersion());
+        if (datasetVersionRecord.getRenameFrom() != null) {
+          restResponseChannel.setHeader(RestUtils.Headers.TARGET_ORIGINAL_NAMED_BLOB_NAME,
```
Comment (justinlin-linkedin) [inline ambry-frontend/src/main/java/com/github/ambry/frontend/GetBlobHandler.java:346]: Could you rename this to "x-ambry-dataset-original-version"?

File: ambry-frontend/src/main/java/com/github/ambry/frontend/GetBlobHandler.java [hunk 6]
```diff
@@ -342,14 +342,23 @@ private String getDatasetVersion(RestRequest restRequest) throws RestServiceExce
...
         // If version is null, use the latest version + 1 from DatasetVersionRecord to construct named blob path.
-        return NAMED_BLOB_PREFIX + SLASH + accountName + SLASH + containerName + SLASH + datasetName + SLASH
-            + datasetVersionRecord.getVersion();
+        if (datasetVersionRecord.getRenameFrom() != null) {
+          return NAMED_BLOB_PREFIX + SLASH + accountName + SLASH + containerName + SLASH + datasetName + SLASH
+              + datasetVersionRecord.getRenameFrom();
+        } else {
+          return NAMED_BLOB_PREFIX + SLASH + accountName + SLASH + containerName + SLASH + datasetName + SLASH
+              + datasetVersionRecord.getVersion();
+        }
```
Comment (justinlin-linkedin) [inline ambry-frontend/src/main/java/com/github/ambry/frontend/GetBlobHandler.java:361]: can we put this logic in a public method of dataset version?

File: ambry-frontend/src/main/java/com/github/ambry/frontend/DeleteBlobHandler.java [hunk 7]
```diff
@@ -222,5 +243,24 @@ private void deleteDatasetVersion(RestRequest restRequest) throws RestServiceExc
             RestServiceErrorCode.getRestServiceErrorCode(ex.getErrorCode()));
       }
     }
+
+
+      private DatasetVersionRecord getDatasetVersionHelper() throws RestServiceException {
```
Comment (justinlin-linkedin) [inline ambry-frontend/src/main/java/com/github/ambry/frontend/DeleteBlobHandler.java:248]: Could we reuse this method in different handlers?

**Review by justinlin-linkedin — APPROVED** [review]

File: ambry-frontend/src/main/java/com/github/ambry/frontend/GetBlobHandler.java [hunk 8]
```diff
@@ -342,14 +342,21 @@ private String getDatasetVersion(RestRequest restRequest) throws RestServiceExce
...
         }
         metrics.getDatasetVersionProcessingTimeInMs.update(System.currentTimeMillis() - startGetDatasetVersionTime);
         // If version is null, use the latest version + 1 from DatasetVersionRecord to construct named blob path.
-        return NAMED_BLOB_PREFIX + SLASH + accountName + SLASH + containerName + SLASH + datasetName + SLASH
-            + datasetVersionRecord.getVersion();
+        if (datasetVersionRecord.getRenameFrom() != null) {
+          return datasetVersionRecord.getRenamedPath(accountName, containerName);
+        } else {
+          return datasetVersionRecord.getOriginalPath(accountName, containerName);
+        }
```
[the remark this answers is not in the document]
Comment (SophieGuo410) [reply ambry-frontend/src/main/java/com/github/ambry/frontend/GetBlobHandler.java]: Sure, updated.

**Pull request merged**